File: java/src/main/java/com/databricks/zerobus/BaseZerobusStream.java
package com.databricks.zerobus;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Base class for Zerobus streams providing common native method declarations and lifecycle
 * management.
 *
 * <p>This class is not intended to be used directly. Use {@link ZerobusJsonStream} for JSON
 * ingestion, {@link ZerobusProtoStream} for Protocol Buffer ingestion, or the deprecated {@link
 * ZerobusStream} for backward compatibility.
 *
 * <h3>Resource Management</h3>
 *
 * <p>Streams hold native resources that are not automatically released by the garbage collector.
 * You <b>must</b> call {@link #close()} when done to avoid native memory leaks. Use
 * try-with-resources for automatic cleanup:
 *
 * <pre>{@code
 * try (ZerobusJsonStream stream = sdk.createJsonStream(...).join()) {
 *     stream.ingestRecordOffset(record);
 * }
 * }</pre>
 *
 * <h3>Thread Safety</h3>
 *
 * <p>Streams are <b>not thread-safe</b>. Each stream instance should be used from a single thread.
 * Do not call methods on the same stream concurrently from multiple threads.
 */
abstract class BaseZerobusStream implements AutoCloseable {

  private static final Logger logger = LoggerFactory.getLogger(BaseZerobusStream.class);

  // Ensure native library is loaded.
  static {
    NativeLoader.ensureLoaded();
  }

  // Native handle to the Rust stream object.
  protected volatile long nativeHandle;

  // Stream properties.
  protected final String tableName;
  protected final StreamConfigurationOptions options;
  protected final boolean isJsonMode;

  // Cached unacked records (populated on close for use in recreateStream).
  protected volatile List<byte[]> cachedUnackedRecords;
  protected volatile List<EncodedBatch> cachedUnackedBatches;

  /**
   * Creates a new BaseZerobusStream.
   *
   * @param nativeHandle the native handle from JNI
   * @param tableName the table name
   * @param options the stream configuration options
   * @param isJsonMode whether this stream is in JSON mode
   */
  protected BaseZerobusStream(
      long nativeHandle, String tableName, StreamConfigurationOptions options, boolean isJsonMode) {
    this.nativeHandle = nativeHandle;
    this.tableName = tableName;
    this.options = options;
    this.isJsonMode = isJsonMode;
  }

  /**
   * Waits for a specific offset to be acknowledged by the server.
   *
   * @param offset the offset to wait for
   * @throws ZerobusException if an error occurs or the wait times out
   */
  public void waitForOffset(long offset) throws ZerobusException {
    ensureOpen();
    nativeWaitForOffset(nativeHandle, offset);
  }

  /**
   * Flushes the stream, waiting for all queued records to be acknowledged.
   *
   * @throws ZerobusException if an error occurs or the flush times out
   */
  public void flush() throws ZerobusException {
    ensureOpen();
    nativeFlush(nativeHandle);
    logger.info("All records have been flushed");
  }

  /**
   * Closes the stream, flushing all pending records first.
   *
   * <p>After closing, unacknowledged records can still be retrieved via subclass methods for use in
   * stream recreation.
   *
   * @throws ZerobusException if an error occurs during close
   */
  @Override
  public void close() throws ZerobusException {
    long handle = nativeHandle;
    if (handle != 0) {
      try {
        // Close the stream first (flushes pending records)
        nativeClose(handle);
      } catch (Exception e) {
        logger.warn("nativeClose failed: {}", e.getMessage());
      }

      // Cache unacked records before destroying the handle (for recreateStream)
      try {
        cachedUnackedRecords = nativeGetUnackedRecords(handle);
        cachedUnackedBatches = nativeGetUnackedBatches(handle);
      } catch (Exception e) {
        logger.warn("Failed to cache unacked records: {}", e.getMessage());
        cachedUnackedRecords = new ArrayList<>();
        cachedUnackedBatches = new ArrayList<>();
      } finally {
        nativeHandle = 0;
        nativeDestroy(handle);
        logger.info("Stream closed");
      }
    }
  }

  /**
   * Returns the cached unacknowledged records as raw bytes.
   *
   * <p>This method returns records that were cached when the stream was closed. It is used
   * internally by recreateStream.
   *
   * @return a list of unacknowledged records as raw bytes, or empty list if none
   */
  protected List<byte[]> getCachedUnackedRecords() {
    return cachedUnackedRecords != null ? cachedUnackedRecords : new ArrayList<>();
  }

  /**
   * Returns the cached unacknowledged batches.
   *
   * <p>This method returns batches that were cached when the stream was closed. It is used
   * internally by recreateStream.
   *
   * @return a list of unacknowledged batches, or empty list if none
   */
  protected List<EncodedBatch> getCachedUnackedBatches() {
    return cachedUnackedBatches != null ? cachedUnackedBatches : new ArrayList<>();
  }

  /**
   * Returns whether the stream is closed.
   *
   * @return true if the stream is closed
   */
  public boolean isClosed() {
    return nativeHandle == 0 || nativeIsClosed(nativeHandle);
  }

  /**
   * Returns the table name for this stream.
   *
   * @return the table name
   */
  public String getTableName() {
    return tableName;
  }

  /**
   * Returns the stream configuration options.
   *
   * @return the stream configuration options
   */
  public StreamConfigurationOptions getOptions() {
    return options;
  }

  /**
   * Ensures the stream is open.
   *
   * @throws ZerobusException if the stream is closed
   */
  protected void ensureOpen() throws ZerobusException {
    if (nativeHandle == 0) {
      throw new ZerobusException("Stream is closed");
    }
    if (nativeIsClosed(nativeHandle)) {
      throw new ZerobusException("Stream is closed");
    }
  }

  // ==================== Native methods implemented in Rust ====================

  private static native void nativeDestroy(long handle);

  protected native CompletableFuture<Void> nativeIngestRecord(
      long handle, byte[] payload, boolean isJson);

  protected native long nativeIngestRecordOffset(long handle, byte[] payload, boolean isJson);

  protected native long nativeIngestRecordsOffset(
      long handle, List<byte[]> payloads, boolean isJson);

  private native void nativeWaitForOffset(long handle, long offset);

  private native void nativeFlush(long handle);

  private native void nativeClose(long handle);

  private native boolean nativeIsClosed(long handle);

  protected native List<byte[]> nativeGetUnackedRecords(long handle);

  protected native List<EncodedBatch> nativeGetUnackedBatches(long handle);
}

File: dotnet/src/Databricks.Zerobus/BaseZerobusStream.cs
using Databricks.Zerobus.Native;
using System.Runtime.InteropServices;

namespace Databricks.Zerobus;

/// <summary>
/// Base class for Zerobus ingestion streams providing common native method
/// declarations, lifecycle management, and offset-based acknowledgment tracking.
/// </summary>
/// <remarks>
/// <para>Streams are thread-safe. Multiple threads can ingest records concurrently;
/// the SDK serializes access to the underlying native handle via a reader-writer lock.</para>
/// <para>Streams implement IDisposable and (on .NET 8+) IAsyncDisposable.
/// Always use <c>using</c> or <c>await using</c> statements to ensure native resources are freed.</para>
/// </remarks>
public abstract class BaseZerobusStream : IDisposable
#if NET8_0_OR_GREATER
    , IAsyncDisposable
#endif

{
    private readonly ReaderWriterLockSlim _lock = new(LockRecursionPolicy.NoRecursion);
    private ZerobusStreamHandle _handle;
    internal int _disposed;

    /// <summary>
    /// The fully qualified Unity Catalog table name.
    /// </summary>
    public string TableName { get; }

    /// <summary>
    /// The stream configuration options.
    /// </summary>
    public StreamConfigurationOptions Options { get; }

    /// <summary>
    /// Whether this is a JSON-mode stream (as opposed to protobuf mode).
    /// </summary>
    protected bool IsJsonMode { get; }

    /// <summary>
    /// Cached unacknowledged records available after stream closure.
    /// </summary>
    protected List<byte[]>? CachedUnackedRecords { get; private set; }

    /// <summary>
    /// Whether the stream has been closed or disposed.
    /// </summary>
    public bool IsClosed => _disposed != 0 || _handle.IsClosed || _handle.IsInvalid;

    internal IntPtr NativeHandle => _handle.DangerousGetHandle();

    /// <summary>
    /// Creates a new base stream with the given native handle and configuration.
    /// </summary>
    internal protected BaseZerobusStream(
        IntPtr nativeHandle,
        string tableName,
        StreamConfigurationOptions options,
        bool isJsonMode)
    {
        _handle = new ZerobusStreamHandle(nativeHandle);
        TableName = tableName ?? throw new ArgumentNullException(nameof(tableName));
        Options = options ?? throw new ArgumentNullException(nameof(options));
        IsJsonMode = isJsonMode;
    }

    // ==================== Lifecycle (sync) ====================

    /// <summary>
    /// Waits until the record at the given offset has been durably acknowledged.
    /// </summary>
    public void WaitForOffset(long offset)
    {
        EnsureOpen();
        WithReadLock(() =>
        {
            CResult result;
            byte ok = NativeMethods.zerobus_stream_wait_for_offset(NativeHandle, offset, out result);
            if (ok == 0)
            {
                string msg = Marshal.PtrToStringAnsi(result.ErrorMessage) ?? "Wait for offset failed";
                SafeFreeErrorMessage(result.ErrorMessage);
                throw new ZerobusException(msg, isRetryable: result.IsRetryable);
            }
        });
    }

    /// <summary>
    /// Blocks until all currently queued records have been durably acknowledged.
    /// </summary>
    public void Flush()
    {
        EnsureOpen();
        WithReadLock(() =>
        {
            byte ok = NativeMethods.zerobus_stream_flush(NativeHandle);
            if (ok == 0)
                throw new ZerobusException("Flush failed. Check logs for details.", isRetryable: true);
        });
    }

    /// <summary>
    /// Gracefully closes the stream, flushing all pending records first.
    /// After calling this method, unacked records can be retrieved via
    /// <see cref="GetCachedUnackedRecords"/>.
    /// </summary>
    public void Close()
    {
        if (_disposed != 0 || _handle.IsClosed || _handle.IsInvalid)
            return;

        WithWriteLock(() =>
        {
            if (_disposed != 0 || _handle.IsClosed || _handle.IsInvalid)
                return;

            byte ok = NativeMethods.zerobus_stream_close(NativeHandle);
            if (ok == 1)
            {
                CacheUnackedData();
            }
        });

        DisposeNativeHandle();
    }

    // ==================== Lifecycle (async) ====================

    /// <inheritdoc cref="WaitForOffset"/>
    public Task WaitForOffsetAsync(long offset)
    {
        return Task.Run(() => WaitForOffset(offset));
    }

    /// <inheritdoc cref="Flush"/>
    public Task FlushAsync()
    {
        return Task.Run(() => Flush());
    }

    /// <inheritdoc cref="Close"/>
    public Task CloseAsync()
    {
        return Task.Run(() => Close());
    }

    // ==================== Unacked records ====================

    /// <summary>
    /// Returns any records that were not acknowledged when the stream was closed.
    /// Returns an empty list if called before the stream is closed.
    /// </summary>
    protected IReadOnlyList<byte[]> GetCachedUnackedRecords()
    {
        return CachedUnackedRecords ?? (IReadOnlyList<byte[]>)Array.Empty<byte[]>();
    }

    /// <summary>
    /// Retrieves unacknowledged records from the live native stream.
    /// </summary>
    protected List<byte[]> GetNativeUnackedRecords()
    {
        EnsureOpen();
        CRecordArray array = NativeMethods.zerobus_stream_get_unacked_records(NativeHandle);
        var result = new List<byte[]>((int)(ulong)array.Len);

        if (array.Records != IntPtr.Zero && array.Len != UIntPtr.Zero)
        {
            int recordSize = Marshal.SizeOf(typeof(CRecord));
            for (ulong i = 0; i < (ulong)array.Len; i++)
            {
                IntPtr recordPtr = array.Records + (int)(i * (ulong)recordSize);
                var record = (CRecord)Marshal.PtrToStructure(recordPtr, typeof(CRecord))!;
                if (record.Data != IntPtr.Zero && record.DataLen != UIntPtr.Zero)
                {
                    int len = (int)(ulong)record.DataLen;
                    var bytes = new byte[len];
                    Marshal.Copy(record.Data, bytes, 0, len);
                    result.Add(bytes);
                }
            }
        }

        NativeMethods.zerobus_free_record_array(array);
        return result;
    }

    /// <summary>
    /// Caches unacknowledged records before the native handle is destroyed.
    /// </summary>
    protected void CacheUnackedData()
    {
        try
        {
            CachedUnackedRecords = GetNativeUnackedRecords();
        }
        catch
        {
            CachedUnackedRecords = new List<byte[]>();
        }
    }

    // ==================== Guard / helpers ====================

    /// <summary>
    /// Throws if the stream is closed or disposed.
    /// </summary>
    protected void EnsureOpen()
    {
        if (IsClosed)
            throw new ZerobusException("Stream is closed or disposed.", isRetryable: false);
    }

    /// <summary>
    /// Executes an action under the read lock. Used for ingest, flush, wait operations.
    /// </summary>
    protected void WithReadLock(Action action)
    {
        _lock.EnterReadLock();
        try { action(); }
        finally { _lock.ExitReadLock(); }
    }

    /// <summary>
    /// Executes an action under the write lock. Used for close/dispose operations.
    /// </summary>
    protected void WithWriteLock(Action action)
    {
        _lock.EnterWriteLock();
        try { action(); }
        finally { _lock.ExitWriteLock(); }
    }

    /// <summary>
    /// Executes a function under the read lock, returning its result.
    /// </summary>
    protected T WithReadLock<T>(Func<T> func)
    {
        _lock.EnterReadLock();
        try { return func(); }
        finally { _lock.ExitReadLock(); }
    }

    /// <summary>
    /// Disposes native resources.
    /// </summary>
    protected void DisposeNativeHandle()
    {
        if (Interlocked.Exchange(ref _disposed, 1) == 0)
        {
            _handle.Dispose();
            _lock.Dispose();
        }
    }

    // ==================== IDisposable / IAsyncDisposable ====================

    /// <summary>
    /// Releases all resources used by the stream.
    /// </summary>
    public void Dispose()
    {
        if (_disposed != 0) return;

        if (!_handle.IsClosed && !_handle.IsInvalid)
        {
            try { CacheUnackedData(); }
            catch { /* best effort */ }
        }

        DisposeNativeHandle();
    }

#if NET8_0_OR_GREATER
    /// <summary>
    /// Asynchronously releases all resources used by the stream.
    /// </summary>
    public ValueTask DisposeAsync()
    {
        Dispose();
        return ValueTask.CompletedTask;
    }
#endif

    private static void SafeFreeErrorMessage(IntPtr msg)
    {
        if (msg != IntPtr.Zero)
        {
            try { NativeMethods.zerobus_free_error_message(msg); }
            catch { /* best effort */ }
        }
    }
}
